### PR TITLE
src/goTestExplorer: implement a test provider for the new test api

### PR DESCRIPTION
**Pull request by firelizzard18**
What this does:
- Implements a test provider using the new API
- Populates the test explorer with a list of modules and/or workspaces
- Populates each module/workspace/package/file as it is expanded
- Creates test entries for the current file, and package/module parents
- Runs tests!

What this does not:
- Debug tests
- Handle stretchr suites

Issues:
- Handling of benchmarks isn't great. But I'm not sure it can get much
  better without changes to `go test`.
- If a test has subtests, I add those subtests. This can get annoying if
  you have a bunch of subtests. Should this be configurable? Disabling
  `testing.followRunningTest` can improve this UX.

Fixes golang/vscode-go#1579

### CODE REVIEW

**Comment by firelizzard18** [comment]
@googlebot I signed it!

File: package.json [hunk 1]
```diff
@@ -70,7 +71,7 @@
     "@types/node": "^13.11.1",
     "@types/semver": "^7.1.0",
     "@types/sinon": "^9.0.0",
-    "@types/vscode": "^1.52.0",
+    "@types/vscode": "^1.57.0",
```
Comment (Psykar) [inline package.json:74]: I needed to bump     "vscode": "^1.52.0" to the same version as this before it compiled for me.

**Review by Psykar — COMMENTED** [review]
I'm currently getting
```
Running the contributed command: 'go.test.cursor' failed.
```

When trying to run tests with this build. Googling isn't helping me all that much with this error unfortunately.

File: package.json (hunk 1 above)
Comment (firelizzard18) [reply package.json:74]: This will be bumped to 1.59 prior to release.

**Comment by firelizzard18** [comment]
@hyangah @suzmue This is ready for review. I'm done fiddling with it - there are some TODOs, and the features I mention in #1641, but I think this is complete enough for a MVP. The API is stable; when 1.59 is released, I'll remove the enable proposed flag, `vscode.d.ts`, and bump the required VSCode version.

**Comment by Psykar** [comment]
Gave this a spin on release of 1.59 today.
Works great mostly!

The "Play" button (red circle in image) at the top gives a popup error "Failed to execute tests"
With the following in the 'Go' output log
`Failed to execute tests: TypeError: request.include is not iterable`

To recover from this requires clearing the popup (which changes the Stop sign back into a Play. But the Play button still doesn't work.

Hitting the button next to the module name itself works fine though (green button in image).

image
image

**Comment by firelizzard18** [comment]
@Psykar That problem should be fixed now

**Comment by Psykar** [comment]
This may be a UI limitation, but test files which have build tags set (which vscode does not have active) show up in the explorer UI as unrun icons (ie. grey circles). 
I suspect the skipped icon (dot with arrow over) might be better UX. (Or ideally a separate icon from both)

**Comment by firelizzard18** [comment]
@Psykar I don't intend to deal with that topic in this MR because it needs more discussion and the MR is already over 2000 lines. I opened #1673 to discuss options.

FYI, if you want to review the PR, you should do it [here](https://go-review.googlesource.com/c/vscode-go/+/330809#message-04daac1e865b511ac0e9f1d1a757953214d5a046). Comments on the GitHub PR are not considered as part of the review process.

**Comment by Psykar** [comment]
> Comments on the GitHub PR are not considered as part of the review process.

No worries - I was steering clear as I wasn't commenting on the code itself, but will direct further feedback to Gerrit.

**Pull request closed without merging**